x-pack/plugin/ml-feature-index-builder/src/main/java/org/elasticsearch/xpack/ml/featureindexbuilder/action/DataFrameJobStateAndStats.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.ml.featureindexbuilder.action;
+
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.xpack.ml.featureindexbuilder.job.FeatureIndexBuilderJobState;
+import org.elasticsearch.xpack.ml.featureindexbuilder.job.DataFrameIndexerJobStats;
+import org.elasticsearch.xpack.ml.featureindexbuilder.job.FeatureIndexBuilderJob;
+
+import java.io.IOException;
+import java.util.Objects;
+
+public class DataFrameJobStateAndStats implements Writeable, ToXContentObject {
+
+    public static final ParseField STATE_FIELD = new ParseField("state");
+    public static final ParseField STATS_FIELD = new ParseField("stats");
+
+    private final String id;
+    private final FeatureIndexBuilderJobState jobState;
+    private final DataFrameIndexerJobStats jobStats;
+
+    public static final ConstructingObjectParser<DataFrameJobStateAndStats, Void> PARSER = new ConstructingObjectParser<>(
+            GetDataFrameJobsAction.NAME,
+            a -> new DataFrameJobStateAndStats((String) a[0], (FeatureIndexBuilderJobState) a[1], (DataFrameIndexerJobStats) a[2]));
+
+    static {
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), FeatureIndexBuilderJob.ID);
+        PARSER.declareObject(ConstructingObjectParser.constructorArg(), FeatureIndexBuilderJobState.PARSER::apply, STATE_FIELD);
+        PARSER.declareObject(ConstructingObjectParser.constructorArg(), (p, c) -> DataFrameIndexerJobStats.fromXContent(p), STATS_FIELD);
+    }
+
+    public DataFrameJobStateAndStats(String id, FeatureIndexBuilderJobState state, DataFrameIndexerJobStats stats) {
+        this.id = id;
```

Comment:
It seems all 3 fields should not be `null`. Add `Objects.requireNotNull` checks? (Btw, just realised the main reason I like those is the way they serve as documentation)